### ISSUE TITLE
Reporting stops when puma plugin is used with phased restarts

### ISSUE THREAD

**Issue by jeffkreeftmeijer** [appsignal/appsignal-ruby#575]
We [received a report](https://app.intercom.io/a/apps/yzor8gyw/inbox/inbox/conversation/24877629510) (private link), explaining the AppSignal integration stops, but doesn't restart when using Puma's phased restarts.

## TODO

- [x] Update docs with new plugin information

**Comment by thijsc** [comment]
We've received another report. The original Puma integration works well with phased restarts. The new plugin added in 2.9.17 does not. Changed this to a bug since it can break if people used the old integration and upgrade.

**Comment by tombruijn** [comment]
We now listen to an `on_booted` event in the new plugin. Maybe there's a phased restart event we can listen too as well?

**Comment by tombruijn** [comment]
https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/conversation/16410700001399

**Comment by tombruijn** [comment]
I tried this out with the following `config/puma.rb` file and commands. I'm unable to reproduce the issue as [it appears in the logs](https://app.intercom.io/a/apps/yzor8gyw/inbox/inbox/conversation/24877629510), but did find another issue with the phased restart.

```ruby
# config/puma.rb
workers 2 # Clustered mode
threads 5, 5

plugin :appsignal

port 3000

app_dir = File.expand_path("../..", __FILE__)
shared_dir = "#{app_dir}/shared"

rails_env = "production"
environment rails_env

# Set up socket location
bind "unix:///#{shared_dir}/sockets/puma.sock"

stdout_redirect "#{shared_dir}/log/puma.stdout.log", "#{shared_dir}/log/puma.stderr.log", true

pidfile "#{shared_dir}/pids/puma.pid"
state_path "#{shared_dir}/pids/puma.state"
activate_control_app
```

```
# In one session
$ bundle exec puma
# In another session
$ bundle exec pumactl phased-restart --control-token foo
```

What I noticed in testing the phased restart is that `prune_bundler` does some things that might break. AppSignal isn't loaded properly during that start. Adding this line to the puma.rb file helps. Without this line the server doesn't respond at all after a phased restart. Instead requests time out. It's not that AppSignal stops reporting data, the whole app breaks.

Adding this line should help with warnings and errors that appear about loading the AppSignal gem multiple times, overwriting constants and such. AppSignal doesn't work if it's loaded more than once.

```ruby
# config/puma.rb
extra_runtime_dependencies ["appsignal"]
prune_bundler
```

**Comment by tombruijn** [comment]
Upon closer inspection it looks like adding the AppSignal plugin breaks the phased restart. I get this in the puma stdout logs:

```
[31781] Early termination of worker
[31785] + Gemfile in context: /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/Gemfile
```

But no error in the stdout or stderr logs or rails logs why it exited. The only other time I see that is when it encounters an error on start.

The server continues to run the old app code because new instances fail on start.

**Comment by tombruijn** [comment]
The error appeared some time later 🤦 

```
/Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler/runtime.rb:313:in `check_for_activated_spec!': You have already activated rack 2.2.3, but your Gemfile requires rack 2.0.6. Prepending `bundle exec` to your command m
ay solve this. (Gem::LoadError)
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler/runtime.rb:31:in `block in setup'
        from /Users/tombruijn/.rubies/ruby-2.6.6/lib/ruby/2.6.0/forwardable.rb:230:in `each'
        from /Users/tombruijn/.rubies/ruby-2.6.6/lib/ruby/2.6.0/forwardable.rb:230:in `each'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler/runtime.rb:26:in `map'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler/runtime.rb:26:in `setup'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler.rb:107:in `setup'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/bundler-1.16.4/lib/bundler/setup.rb:20:in `<top (required)>'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/boot.rb:3:in `require'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/boot.rb:3:in `<top (required)>'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/application.rb:1:in `require_relative'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/application.rb:1:in `<top (required)>'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/environment.rb:2:in `require_relative'
        from /Users/tombruijn/appsignal/examples/apps/rails-5+action-cable/config/environment.rb:2:in `<top (required)>'
        from config.ru:3:in `require_relative'
        from config.ru:3:in `block in <main>'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/rack/builder.rb:190:in `instance_eval'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/rack/builder.rb:190:in `initialize'
        from config.ru:in `new'
        from config.ru:in `<main>'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/rack/builder.rb:176:in `eval'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/rack/builder.rb:176:in `new_from_string'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/rack/builder.rb:167:in `parse_file'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/configuration.rb:321:in `load_rackup'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/configuration.rb:246:in `app'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/runner.rb:165:in `app'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/runner.rb:172:in `start_server'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:269:in `worker'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:138:in `block (2 levels) in spawn_workers'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:138:in `fork'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:138:in `block in spawn_workers'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:134:in `times'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:134:in `spawn_workers'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:204:in `check_workers'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cluster.rb:487:in `run'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/launcher.rb:172:in `run'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/lib/puma/cli.rb:80:in `run'
        from /Users/tombruijn/.gem/ruby/2.6.6/gems/puma-4.3.5/bin/puma-wild:31:in `<main>'
```

Anyway, updating the `extra_runtime_dependencies` to include `rack`, a AppSignal gem dependency` worked:

```ruby
# config/puma.rb
extra_runtime_dependencies ["appsignal", "rack"]
prune_bundler
```

**Comment by tombruijn** [comment]
Potential fix to ship in the gem:

```diff
diff --git lib/puma/plugin/appsignal.rb lib/puma/plugin/appsignal.rb
index 811a2ced..4f5e5164 100644
--- lib/puma/plugin/appsignal.rb
+++ lib/puma/plugin/appsignal.rb
@@ -13,12 +14,29 @@ APPSIGNAL_PUMA_PLUGIN_LOADED = true
 # For even more information:
 # https://docs.appsignal.com/ruby/integrations/puma.html
 Puma::Plugin.create do
+  def config(c)
+    c.extra_runtime_dependencies ["appsignal", "rack"]
+  end
```

**Comment by tombruijn** [comment]
Actually, can't use the `config` method as that resets the user defined config. Which we still need, because our plugin is not going to set it.

But also, adding the `extra_runtime_dependencies` loads AppSignal and Rack in the main process, so when a gem gets updated or something they won't update, staying locked at the original version.

Also, because we initialize the config, the config doesn't get updated if new workers start during a phased restart, because they'll use the config from the main process instead. Resetting the config on worker boot with `Appsignal.config = nil` doesn't seem something we'd want to do either, and wouldn't solve the gem version issue.

```ruby
# config/puma
on_worker_boot do
  Appsignal.config = nil # Don't do this, example of a bad way to fix half the problem
end
```

So it's back to the drawing board. We shouldn't load AppSignal in the puma main process if it can't reload the gem version and config.

Maybe we should put the puma probe back in the worker processes, instead of the main process. It will report the data multiple times, but as long as it only reports gauges it should be fine. It wouldn't be able to report counters, because that would be reported X times where X is the amount of workers. Or find a way to only run the probe once, in one of the workers, which is what we tried to do by running it in the main process.

`Puma.stats` won't return anything useful in clustered mode inside a worker, we need another way to fetch the data from the main process. Maybe call the controller app's API endpoint and fetch the data that way? Maybe puma exposes it that way.


TLDR: We can't support phased restarts this way.

**Comment by tombruijn** [comment]
We should probably document on the puma page that we do not support phased restarts and the plugin will break the app, so it's a DO NOT USE scenario.

**Comment by matsimitsu** [comment]
We are probably safer running a job in our `minutely probe` that fetches and parses the state file exposed by Puma. 

This way we don't have to be "anywhere" and customers can configure the required parameters (state path/token) in the initialiser of the probe. 

What do you think @tombruijn ?

**Comment by tombruijn** [comment]
We can't reliably get it for every scenario that "just works" from the puma processes, main or workers. So yeah we need to fetch it some other way. Your PR #655 looks like the right direction.

**Comment by shairyar** [comment]
https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/4356044/conversations/16410700022903

**Comment by shairyar** [comment]
https://app.intercom.com/a/apps/yzor8gyw/inbox/inbox/4356044/conversations/16410700034169

**Comment by tombruijn** [comment]
Let's add a test setup to the test-setups repo

**Comment by tombruijn** [comment]
Test setup added here: https://github.com/appsignal/test-setups/tree/e74ed9101ad7f8b5af52d5f6a3ccc439712eb3a2/ruby/sinatra-puma

**Comment by tombruijn** [comment]
I created a PR with a potential fix: https://github.com/appsignal/appsignal-ruby/pull/732

**Comment by tombruijn** [comment]
I've released Ruby gem 3.0.7 with a potential fix.

@shairyar can you ask the users to try out the new release with phased restarting?

**Comment by tombruijn** [comment]
We also updated the docs page yesterday: https://docs.appsignal.com/ruby/integrations/puma.html

**Comment by tombruijn** [comment]
I'm closing this issue. We can reopen when people come back with reports.
